### PR TITLE
252 add an ability for the unicorn

### PR DESCRIPTION
**Pull request by Monkey-bored**
This was harder than expected.
I've done some stuff here knowing there might but a better solution but it's better to give a solution and improve it then not give any solution at all.

### CODE REVIEW

File: core/development/LogicAdapter.ts [hunk 1]
```diff
@@ -408,3 +416,27 @@ export function buyItem(itemId: string) {
...
+  const targetPiece = getPieceByElement(targetPieceElement, boardId);
+  if (!targetPiece) return;
+
+  const targetablePieces = unicornPiece.getAttackablePieces();
+
+  // If the piece is not in the array, return
+  if (targetablePieces.filter(piece => piece === targetPiece).length === 0) return;
+  unicornPiece.player.usedAbility = true;
+  handleOverworldKill(targetPiece);
+}
```
Comment (Ido-Barnea) [inline core/development/LogicAdapter.ts:442]: missing line at EOF

File: core/development/LogicAdapter.ts [hunk 2]
```diff
@@ -408,3 +416,27 @@ export function buyItem(itemId: string) {
...
+  const unicornPiece = getPieceByElement(pieceElement, boardId);
+  if (!(unicornPiece instanceof Unicorn) || unicornPiece.player.usedAbility) return;
+
+  const targetPiece = getPieceByElement(targetPieceElement, boardId);
+  if (!targetPiece) return;
+
+  const targetablePieces = unicornPiece.getAttackablePieces();
+
+  // If the piece is not in the array, return
+  if (targetablePieces.filter(piece => piece === targetPiece).length === 0) return;
```
Comment (Ido-Barnea) [inline core/development/LogicAdapter.ts:439]: If you have to explain the code using a comment, it's not readable enough

File: core/development/logic/pieces/Unicorn.ts [hunk 3]
```diff
@@ -10,6 +11,42 @@ export class Unicorn extends Piece {
...
+    const attackPositionsOffsets = [
+      { deltaX: 1, deltaY: 0 },
+      { deltaX: 0, deltaY: 1 },
+      { deltaX: 1, deltaY: 1 },
+      { deltaX: -1, deltaY: 0 },
+      { deltaX: 0, deltaY: -1 },
+      { deltaX: -1, deltaY: -1 },
+      { deltaX: -1, deltaY: 1 },
+      { deltaX: 1, deltaY: -1},
+    ]
```
Comment (Ido-Barnea) [inline core/development/logic/pieces/Unicorn.ts:30]: If you have to explain the code using a comment, it's not readable enough

File: core/development/ui/Events.ts [hunk 4]
```diff
@@ -15,11 +16,16 @@ import {
...
 import { HEAVEN_BOARD, HELL_BOARD, OVERWORLD_BOARD } from './BoardManager';
+import { hideUnicornAttackButton } from './Screen';
 
 const MOVEMENT_TO_CLICK_THRESHOLD = 10;
 
 let draggedElement: HTMLElement | undefined;
 
+let selectedPieceElement: HTMLElement | undefined = undefined;
+
+let attackingUnicorn: HTMLElement | undefined = undefined;
```
Comment (Ido-Barnea) [inline core/development/ui/Events.ts:27]: Isn't the default value already `undefined`?

File: core/development/ui/Events.ts [hunk 5]
```diff
@@ -40,6 +46,8 @@ const SHOP_UPGRADE_SWAPPER = document.getElementById('shop-upgrade-swapper');
 const SHOP_CONTAINER = document.getElementById('shop-container');
 const UPGRADES_CONTAINER = document.getElementById('piece-upgrades-container');
 
+const UNICORN_ATTACK_BUTTON = document.getElementById('unicorn-attack')
```
Comment (Ido-Barnea) [inline core/development/ui/Events.ts:49]: Missing `;`

File: core/development/ui/Events.ts [hunk 6]
```diff
@@ -235,7 +244,14 @@ function onPieceClick(pieceElement: HTMLElement) {
   while (!boardElement?.classList.contains('board')) {
     boardElement = boardElement?.parentElement ?? undefined;
   }
+  if (attackingUnicorn) {
+      unicornAttackAttempt(attackingUnicorn, pieceElement, boardElement.id);
+      selectedPieceElement = attackingUnicorn = undefined;
```
Comment (Ido-Barnea) [inline core/development/ui/Events.ts:249]: Not really readable, please separate to multiple lines.

File: core/development/ui/Events.ts [hunk 7]
```diff
@@ -235,7 +244,14 @@ function onPieceClick(pieceElement: HTMLElement) {
...
+      hideUnicornAttackButton();
+      return;
+  }
 
+  selectedPieceElement = pieceElement;
   onPieceSelected(pieceElement, boardElement.id);
```
Comment (Ido-Barnea) [inline core/development/ui/Events.ts:255]: `if` + `else` would probably be more readable than a `return;`

File: core/development/ui/Screen.ts [hunk 8]
```diff
@@ -33,3 +36,18 @@ export function showWinningAlert(winner: string) {
...
+
+export function hideUnicornAttackButton() {
+  if (!unicornAttackButton) return;
+  unicornAttackButton.classList.add('collapsed');
+}
+
+export function showUnicornAttackButton() {
+  if (!unicornAttackButton) return;
+  unicornAttackButton.classList.remove('collapsed');
+}
```
Comment (Ido-Barnea) [inline core/development/ui/Screen.ts:53]: Missing line at EOF

**Review by Ido-Barnea — CHANGES_REQUESTED** [review]
Most of my notes are about readability so please read them carefully and let me know if there is anything you don't understand. Other than that, great job 👍🏼 💯

File: core/development/LogicAdapter.ts (hunk 1 above)
Comment (Monkey-bored) [reply core/development/LogicAdapter.ts:442]: Rookie mistake

File: core/development/logic/pieces/Unicorn.ts (hunk 3 above)
Comment (Monkey-bored) [reply core/development/logic/pieces/Unicorn.ts:30]: I was being consistent with the code below in the Unicorn class
``
    // Define all possible unicorn move offsets
    const unicornMoveOffsets = [
      { deltaX: 2, deltaY: 1 },
      { deltaX: 2, deltaY: -1 },
      { deltaX: -2, deltaY: 1 },
      { deltaX: -2, deltaY: -1 },
      { deltaX: 1, deltaY: 2 },
      { deltaX: 1, deltaY: -2 },
      { deltaX: -1, deltaY: 2 },
      { deltaX: -1, deltaY: -2 },
    ];
``

Comment (Monkey-bored) [reply core/development/logic/pieces/Unicorn.ts:30]: I hate github
image

File: core/development/ui/Events.ts (hunk 4 above)
Comment (Monkey-bored) [reply core/development/ui/Events.ts:27]: Yeah, and?

File: core/development/ui/Events.ts (hunk 5 above)
Comment (Monkey-bored) [reply core/development/ui/Events.ts:49]: Thats what 8 hours of python a day do to you

File: core/development/logic/pieces/Unicorn.ts (hunk 3 above)
Comment (Ido-Barnea) [reply core/development/logic/pieces/Unicorn.ts:30]: I never said I wrote it well in the first place 😆 😢

File: core/development/ui/Events.ts (hunk 4 above)
Comment (Ido-Barnea) [reply core/development/ui/Events.ts:27]: Why set it to `undefind`?

File: core/development/logic/pieces/Unicorn.ts (hunk 3 above)
Comment (Monkey-bored) [reply core/development/logic/pieces/Unicorn.ts:30]: Should I just remove the comment? I think the name pretty much explains it

File: core/development/ui/Events.ts (hunk 4 above)
Comment (Monkey-bored) [reply core/development/ui/Events.ts:27]: Cuz im quirky

File: core/development/logic/pieces/Unicorn.ts (hunk 3 above)
Comment (Ido-Barnea) [reply core/development/logic/pieces/Unicorn.ts]: sure

File: core/development/LogicAdapter.ts [hunk 9]
```diff
@@ -428,15 +428,21 @@ export function getPieceByElement(pieceElement: HTMLElement, boardId: string) {
...
   const targetPiece = getPieceByElement(targetPieceElement, boardId);
   if (!targetPiece) return;
 
   const targetablePieces = unicornPiece.getAttackablePieces();
 
-  // If the piece is not in the array, return
-  if (targetablePieces.filter(piece => piece === targetPiece).length === 0) return;
+  console.log(targetablePieces);
+  console.log(targetPiece);
+  console.log(targetablePieces.includes(targetPiece));
```
Comment (Ido-Barnea) [inline core/development/LogicAdapter.ts:443]: `console.log()` in production is a big "no no"

**Review by Ido-Barnea — CHANGES_REQUESTED** [review]

File: core/development/LogicAdapter.ts (hunk 9 above)
Comment (Monkey-bored) [reply core/development/LogicAdapter.ts]: What console.log()

**Review by Ido-Barnea — APPROVED** [review]
LGTM ✌🏼 🛩️ 😎

**Pull request merged**